### ISSUE TITLE
Automated headless unit testing

### ISSUE THREAD

**Issue by epost** [CategoricalData/fql#121]
Automated unit testing seems like a good thing to have, especially in conjunction with automated builds and testing by some CI (Continuous Integration) service. I am creating this issue partly as a point of reference for some PR's I intend to make.

**Comment by peterygates** [comment]
I am a big fan of automated unit testing. If you are willing to create a branch and some tests I would support that. I think we might want to have a broader conversation about how to proceed. Let me know if you can schedule some time to have a call.

**Comment by epost** [comment]
Hi Peter, thanks for your response! :) I already made a Pull Request; please see the #122 link. If you could have a look at that, I'll try and sketch some of my ideas for this so I can get back to you re the call ASAP.

**Comment by wisnesky** [comment]
There is a "run all examples" item in the aql menu bar or nearby.  It's
what I've been using as a "unit test".

**Comment by epost** [comment]
Hi @wisnesky, thanks for your response! I'd noticed the GUI-based test suite, and I took some inspiration from it in producing #122. However, when it comes to enabling AQL usage and contributions by others, and QA (e.g. helping to ensure non-broken releases) it would seem like a good idea to follow the standard practice of being able to run standardised, headless unit tests. That's why I issued #122 as a starting point. (NB: It should be quite easy to refactor both the GUI and the test code to eliminate duplicate code.)

**Comment by epost** [comment]
@peterygates Btw, #122 seems to me to be a good idea regardless of any specific choices of how to proceed with automated testing, since it simply adds a very basic test (i.e. the pre-existing test suite that Ryan mentioned) in a standardised format (Maven, JUnit) to a build file that we *already have*, i.e. the Maven `pom.xml`.

**Comment by peterygates** [comment]
All,

I have had the notion of building a set of AQL modules that attempts to
achieve some degree of test coverage. Ryan's examples are a good start.
With all of the other priorities I have just not made much progress on
this. Seems like a review of the reference manual might be a good starting
place for defining test coverage. I know of Ryan's thoughts about
attempting to create an exhaustive test suite, but I still see value in
thinking about a framework that focuses on a risk based approach.

What is your wisdom regarding coding metrics to identify risk? Years ago I
spent some time researching this stuff, but I am sure what I know is out of
date. For example McCabe's cyclomatic complexity
<http://www.literateprogramming.com/mccabe.pdf> was something I tried to
understand.

**Comment by peterygates** [comment]
@epost, love your  enthusiasm. Heads up, I am going to be on the road and
largely unavailable until Thursday 19, so may be a while before I catch up
on this exchange.

**Comment by wisnesky** [comment]
I agree, those should not be in the default tests, although it is probably possible to change CSV.aql to use a Java ‘temporary file’ rather than a file hardcoded to my particular development environment.

**Comment by wisnesky** [comment]
Hi Erik, Fred, and Peter,

I just wanted to check that you all have enough GitHub ‘privileges’ or equivalent to contribute to AQL without my oversight, because I don’t want to be a single point of failure or latency or bandwidth.  So do please let me know if there’s anything you need to enable you to act independently on GitHub.  So far it seems like things are going smoothly because I don’t see any open ‘pull requests’.

Ryan

**Comment by phreed** [comment]
Ryan,
Presently I do not have write authority to the repository.
Would it be good to add a work plan somewhere?
The wiki maybe?
Thanks
Fred

**Comment by phreed** [comment]
https://github.com/orgs/CategoricalData/projects

**Comment by wisnesky** [comment]
Can you tell me how to give you authority to write to the repository?

**Comment by epost** [comment]
Hi folks, I moved these topics into their own issues so we can keep this one on topic, and close the issues when they're done.

**Comment by phreed** [comment]
Back on topic.
It seems we have three kinds of examples.
- unit tests
- illustrative examples
- domain specific examples

The **domain specific examples** should go in their own projects.
http://catinf.com/wiki/Main_Page
https://github.com/CategoricalData/fql/tree/master/contrib/AQL_Examples
I would like to see some stock databases included in this group.

  - https://www.w3schools.com/sql/default.asp
  - https://www.ibm.com/support/knowledgecenter/en/SSEPEK_10.0.0/sqlref/src/tpc/db2z_sql_subselectexamples.html

The **illustrative examples** show up in the fql-app and demonstrate common usage.
Some of these examples make use of significant amounts of data or interaction with a data base manager.

The **unit tests** exercise the capablity of the system and should probably be written as property based tests, https://en.wikipedia.org/wiki/QuickCheck

The test suites would be aggregates of these examples.
I think the Tutorial\ TSP.aql should be moved into its own "strictly illustrative" group so it can be excluded from the unit test suite.
The CSV.aql can be provided with data files in the project.

What say ye?

**Comment by wisnesky** [comment]
Works for me.  On a related note, the java code to read examples out of the aql.jar file is extremely delicate.  Maybe this is a good time to address that as well?
